service/raft/group0_state_machine.cc

```diff
@@ -43,11 +43,34 @@
 #include "db/config.hh"
 #include "replica/database.hh"
 #include "service/raft/group0_state_machine_merger.hh"
+#include "gms/feature_service.hh"
 
 namespace service {
 
 static logging::logger slogger("group0_raft_sm");
 
+group0_state_machine::group0_state_machine(raft_group0_client& client, migration_manager& mm, storage_proxy& sp, storage_service& ss, raft_address_map& address_map, gms::feature_service& feat, bool topology_change_enabled)
+        : _client(client), _mm(mm), _sp(sp), _ss(ss), _address_map(address_map), _topology_change_enabled(topology_change_enabled)
+        , _topology_on_raft_support_listener(feat.supports_consistent_topology_changes.when_enabled([this] () noexcept {
+            // Using features to decide whether to start fetching topology snapshots
+            // or not is technically not correct because we also use features to guard
+            // whether upgrade can be started, and upgrade starts by writing
+            // to the system.topology table (namely, to the `upgrade_state` column).
+            // If some node at that point didn't mark the feature as enabled
+            // locally, there is a risk that it might try to pull a snapshot
+            // and will decide not to use `raft_pull_topology_snapshot` verb.
+            //
+            // The above issue is mitigated by requiring administrators to
+            // wait for the duration of ring delay before starting the upgrade
```

Comment:
They should verify the feature is enabled (waiting as long as it takes), not wait for ring_delay.